### PR TITLE
[ROS] Crystal

### PR DESCRIPTION
**Pull request by ruffsl**
Support for ROS2 Crystal
https://github.com/ros2/ros2/issues/529

TODO list upstreaming to Docker Hub Official Library:

* [x] Update docker templates for ROS2:
  * See https://github.com/osrf/docker_templates/pull/46
* [x] Update PR to version pin release
  * Set version in template to empty to force remote lookup
* [x] Create Upstream Documentation PR
  * https://github.com/docker-library/docs/blob/master/README.md
  * See https://github.com/docker-library/docs/pull/1381
* [x] Create Upstream Library PR
  * https://github.com/docker-library/official-images/blob/master/NEW-IMAGE-CHECKLIST.md
  * See https://github.com/docker-library/official-images/pull/5183

### CODE REVIEW

File: ros2/manifest.yaml [hunk 1]
```diff
@@ -47,19 +52,52 @@ release_names:
...
+                                    - "$release_name-ros1-bridge-$os_code_name"
+    crystal:
+        eol: 2020-XX
+        os_names:
+            ubuntu:
+                os_code_names:
+                    bionic:
+                        rosdistro_name: melodic
+                        <<: *DEFAULT
+                        archs:
```
Comment (ruffsl) [inline ros2/manifest.yaml]: Would be nice to target arm32v7 as well, but I'm guessing that's not on the buildfarm roadmap?

File: ros2/crystal/ubuntu/bionic/ros1-bridge/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+# generated from docker_images/create_ros2_ros1_bridge_image.Dockerfile.em
+FROM osrf/ros2:crystal-desktop
+
+# setup keys
+RUN apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv-keys 421C365BD9FF1F717815A3895523BAEEB01FA116
+
+# setup sources.list
+RUN echo "deb http://packages.ros.org/ros/ubuntu `lsb_release -sc` main" > /etc/apt/sources.list.d/ros-latest.list
+
+ENV ROS1_DISTRO melodic
```
[the remark this answers is not in the document]
Comment (sloretz) [reply ros2/crystal/ubuntu/bionic/ros1-bridge/Dockerfile]: Yes. The bridge needs the ROS 1 and ROS 2 distro to be on the same ubuntu release. Crystal debs will be for bionic so the bridge must be built against melodic.

IIUC Crystal can be built from source on xenial so someone could bridge kinetic to it, but I don't think that's relevant for this dockerfile.

File: ros2/manifest.yaml [hunk 3]
```diff
@@ -47,19 +52,52 @@ release_names:
...
+                            - amd64
+                            - arm64v8
+                        tag_names:
```
[the remark this answers is not in the document]
Comment (sloretz) [reply ros2/manifest.yaml]: I don't think so. IIUC new meta packages would first show up in https://github.com/ros2/variants

File: ros2/manifest.yaml (hunk 1 above)
Comment (sloretz) [reply ros2/manifest.yaml]: @nuclearsandwich says `It's Complicated [...] at release the only planned architectures for crystal are arm64 and amd64`

File: ros2/manifest.yaml [hunk 4]
```diff
@@ -47,19 +52,52 @@ release_names:
...
                             desktop:
                                 aliases:
                                     - "$release_name-desktop"
+                                    - "$release_name-desktop-$os_code_name"
                             ros1-bridge:
                                 aliases:
                                     - "$release_name-ros1-bridge"
+                                    - "$release_name-ros1-bridge-$os_code_name"
+    crystal:
+        eol: 2020-XX
```
[the remark this answers is not in the document]
Comment (sloretz) [reply ros2/manifest.yaml]: I think the current plan is 1 year of support ending December 14th, 2019

[the remark this answers is not in the document]
Comment (sloretz) [reply ros2/manifest.yaml]: The PR for the target platforms REP says `TBD, at least December 2019` https://github.com/ros-infrastructure/rep/pull/177/files

File: ros2/manifest.yaml (hunk 1 above)
Comment (ruffsl) [reply ros2/manifest.yaml]: ... I can dream. @nuclearsandwich , if there is anything I could do to help in advancing arm32v7, like debug tests or investigate issues, I gladly give it go. There are a lot of hobbyist and academic users that use armhf, and if we had readily available images they could easily pull from, I think it would really help bolster ROS2 exposure and adoption.

Comment (nuclearsandwich) [reply ros2/manifest.yaml]: @ruffsl I don't want to sidetrack this discussion too much, there's a lot that I think can be done. For one, I'm not sure if anyone has done a from-source build of ROS 2 on an armv7 system recently. If it's something you're interested in I'd suggest a discourse thread.

File: ros2/.config/platform.yaml.em [hunk 5]
```diff
@@ -10,4 +10,4 @@ platform:
     maintainer_name:
     arch: amd64
     type: distribution
-    version: false
+    version:
```
Comment (ruffsl) [inline ros2/.config/platform.yaml.em]: Update Dockerfile upon release sync to pin package versions.

File: ros2/crystal/ubuntu/bionic/ros-core/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+# setup sources.list
+RUN . /etc/os-release \
+    && echo "deb http://repo.ros2.org/$ID/main `lsb_release -sc` main" > /etc/apt/sources.list.d/ros2-latest.list
+
+# setup environment
+ENV LANG C.UTF-8
+ENV LC_ALL C.UTF-8
+
+ENV ROS_DISTRO crystal
```
Comment (ruffsl) [inline ros2/crystal/ubuntu/bionic/ros-core/Dockerfile]: Do you think we should install and init rosdep in the core ros2 image, like we do for ros1? E.g:
```
# install bootstrap tools
RUN apt-get update && apt-get install --no-install-recommends -y \
    python3-rosdep \
    python3-rosinstall \
    python3-vcstools \
    && rm -rf /var/lib/apt/lists/*

# bootstrap rosdep
ENV ROSDISTRO_INDEX_URL https://raw.githubusercontent.com/ros2/rosdistro/ros2/index.yaml
RUN rosdep init \
    && rosdep update
```
Also, do we still need to set the `ROSDISTRO_INDEX_URL`? Ping @tfoote

Comment (nuclearsandwich) [reply ros2/crystal/ubuntu/bionic/ros-core/Dockerfile]: > Do you think we should install and init rosdep in the core ros2 image, like we do for ros1?

I don't have a considered opinion on this but if we're doing it in the ROS images it's reasonable for end users to expect it in the ROS 2 images as well. ROS 2 has used the same rosdep database as ROS since at least Bouncy and maybe even Ardent, my memory fails me here. So adding the rosdep initialization should just be a net positive.



> Also, do we still need to set the `ROSDISTRO_INDEX_URL`?

ROS 2 is still using a forked rosdistro index in https://github.com/ros2/rosdistro. The documentation for [releasing a ROS 2 package with Bloom](https://index.ros.org/doc/ros2/Releasing-a-ROS-2-package-with-bloom/#differences-from-ros-1-bloom) will always have the latest information but for Crystal the ROSDISTRO_INDEX_URL should be set to
```
https://raw.githubusercontent.com/ros2/rosdistro/ros2/index-v4.yaml
```

Comment (ruffsl) [reply ros2/crystal/ubuntu/bionic/ros-core/Dockerfile]: Is this `ROSDISTRO_INDEX_URL` expected to bump version often, e.g. I should parameterize it into the yaml config rather than leave it in the template.em? Can ardent/bounce/crystal all use the same index url from above? Will this index be roll back upstream from the `rosdistro` is forked, or indefinitely remain separate/independent?

Comment (nuclearsandwich) [reply ros2/crystal/ubuntu/bionic/ros-core/Dockerfile]: > Is this ROSDISTRO_INDEX_URL expected to bump version often, e.g. I should parameterize it into the yaml config rather than leave it in the template.em?

Hopefully this should be a highly infrequent event. The change is related to https://github.com/ros-infrastructure/rep/pull/181 for other images, the default rosdistro index url inside rosdep has already been updated to use the v4 index in ros/rosdistro.

> Can ardent/bounce/crystal all use the same index url from above?

Yep.

Comment (ruffsl) [reply ros2/crystal/ubuntu/bionic/ros-core/Dockerfile]: Thanks, PR updated to incorporate bootstrapping tools.

**Comment by nuclearsandwich** [comment]
We should set up these images to use the new ROS2 main repository. The documentation hasn't switched over yet but the repositories are live. https://github.com/ros2/ros2_documentation/pull/59

File: README.md [hunk 7]
```diff
@@ -4,6 +4,14 @@ This repo hosts the Dockerfiles used to generate the official images for OSRF pr
 ##### Dockerfile CI [![Build Status](https://travis-ci.org/osrf/docker_images.svg?branch=master)](https://travis-ci.org/osrf/docker_images)
 
 ## Official Library
+### [ROS2 ![Docker Pulls](https://img.shields.io/docker/pulls/_/ros2.svg) ![Docker Stars](https://img.shields.io/docker/stars/_/ros2.svg)](https://registry.hub.docker.com/_/ros2/)
```
Comment (sloretz) [inline README.md]: It looks like ros2 doesn't exist on the official docker hub. Is the goal here to make the official images match the split in `osrf/ros` and `osrf/ros2`?

**Comment by ruffsl** [comment]
> We should set up these images to use the new ROS2 main repository. The documentation hasn't switched over yet but the repositories are live. ros2/ros2_documentation#59

@nuclearsandwich , please see 1e3e24d and https://github.com/osrf/docker_templates/pull/46/commits/fcce956f76e0899f10d9d194f1e2a9e2535f85fd

> In the templates an $ID variable is used for main or testing. That logic needs to be updated to select between http://packages.ros.org/ros2/ubuntu and http://packages.ros.org/ros2-testing/ubuntu

I'll try and expand the template to account for this later on, though it shouldn't block anything here.

File: README.md (hunk 7 above)
Comment (ruffsl) [reply README.md]: I've now reworked the PRs to consolidate the versions into just the one official ros repo.

**Comment by ruffsl** [comment]
@sloretz @tfoote and @nuclearsandwich , I've modified the PR in accordance with the existing ROS repo in the docker library. It now shares the same entrypoint and repo structure. I've also update the upstream PR to the library to point to the latest state of this PR. Let me know if you have any more suggestions or not.

File: README.md [hunk 8]
```diff
@@ -40,12 +40,6 @@ This repo hosts the Dockerfiles used to generate the official images for OSRF pr
 * amd64 [![Docker Build Status](https://img.shields.io/docker/build/osrf/gazebo.svg)](https://hub.docker.com/r/osrf/gazebo/builds/)
 
 
-### [SROS ![Docker Pulls](https://img.shields.io/docker/pulls/osrf/sros.svg) ![Docker Stars](https://img.shields.io/docker/stars/osrf/sros.svg)](https://registry.hub.docker.com/u/osrf/sros/)
```
Comment (sloretz) [inline README.md]: Is removing these badges from the README accidental? It seems like an unrelated change.

**Comment by sloretz** [comment]
It looks like an exception is raised on my machine when creating the dockerfiles for `ros/source/ros2`. Is there a different branch of `osrf/docker_templates` I should be using?

```
$ ./create_dockerfiles.py dir -d ./source/ros2
RuntimeError processing template 'docker_images/create_ros2_source_image.Dockerfile.em'
Traceback (most recent call last):
  File "./create_dockerfiles.py", line 70, in <module>
    main()
  File "./create_dockerfiles.py", line 67, in main
    create_files(data)
  File "/home/sloretz/projects/osrf-docker/env/lib/python3.5/site-packages/docker_templates-0.0.1-py3.5.egg/docker_templates/create.py", line 45, in create_files
    create_dockerfile(template_name, data, dockerfile_dir, verbose)
  File "/home/sloretz/projects/osrf-docker/env/lib/python3.5/site-packages/ros_buildfarm-2.0.1-py3.5.egg/ros_buildfarm/templates/__init__.py", line 162, in create_dockerfile
    content = expand_template(template_name, data)
  File "/home/sloretz/projects/osrf-docker/env/lib/python3.5/site-packages/ros_buildfarm-2.0.1-py3.5.egg/ros_buildfarm/templates/__init__.py", line 79, in expand_template
    template_path = get_template_path(template_name)
  File "/home/sloretz/projects/osrf-docker/env/lib/python3.5/site-packages/ros_buildfarm-2.0.1-py3.5.egg/ros_buildfarm/templates/__init__.py", line 40, in get_template_path
    raise RuntimeError("Failed to find template '%s'" % template_name)
RuntimeError: Failed to find template 'docker_images/create_ros2_source_image.Dockerfile.em'
```

File: README.md (hunk 8 above)
Comment (ruffsl) [reply README.md]: It was intentional. With official Ros 2 docker images offering security features, I see little need to distract new users from this with older unfinished security approaches.

**Comment by ruffsl** [comment]
The ROS 2 source docker file hasn't been generated for some time. I'm not surprised something broke, but this is not a blocker for the office images. Looks like the path to the em file just needs to be updated to reflect the directory restructure in the template package. Once I fix this, is there any other last changes to the official Docker images? Otherwise continue with the upstream process.

**Comment by sloretz** [comment]
>  Once I fix this, is there any other last changes to the official Docker images? Otherwise continue with the upstream process.

No changes; it looks good to me.
